src/KDMqtt/mqtt.h

```diff
@@ -0,0 +1,250 @@
+    virtual bool isValidTopicNameForSubscription(const std::string &topic) = 0;
+};
+
+/*
+ * Class: MqttLib
+ *
+ * This class exposes mosquitto library functions to
+ * the application / business logic.
+ */
+class KDMQTT_API MqttLib : public IMqttLib
+{
+    friend class MqttClient;
+    friend class MqttUnitTestHarness;
+
+private:
+    MqttLib();
+    ~MqttLib() = default;
+
+public:
+    static MqttLib &instance();
+
+    int init() override;
+    int cleanup() override;
+
+    [[nodiscard]] bool isInitialized() const override;
+    bool isValidTopicNameForSubscription(const std::string &topic) override;
+
+protected:
+    int version(int *major, int *minor, int *revision);
+
+    bool checkMosquittoResultAndDoDebugPrints(int result, std::string_view func = {});
+
+    std::string_view connackString(int connackCode);
+    std::string_view errorString(int errorCode);
+    std::string_view reasonString(int reasonCode);
+
+private:
+    MosquittoLib *m_mosquittoLib;
+    bool m_isInitialized;
+};
+
+/*
+ * Class: IMqttClient
+ *
+ * This is an abstract class specifing a generic
+ * interface exposed to application / business logic
+ * to access MQTT client implementations.
+ */
+class KDMQTT_API IMqttClient
+{
+public:
+    enum class ConnectionState {
+        CONNECTING,
+        CONNECTED,
+        DISCONNECTING,
+        DISCONNECTED
+    };
+
+    enum class SubscriptionState {
+        SUBSCRIBING,
+        SUBSCRIBED,
+        UNSUBSCRIBING,
+        UNSUBSCRIBED
+    };
+
+    KDBindings::Property<ConnectionState> connectionState{ ConnectionState::DISCONNECTED };
+    KDBindings::Property<SubscriptionState> subscriptionState{ SubscriptionState::UNSUBSCRIBED };
+
+    KDBindings::Property<std::vector<std::string>> subscriptions{};
+
+    KDBindings::Signal<int /*msgId*/> msgPublished;
+    KDBindings::Signal<const mosquitto_message * /*msg*/> msgReceived;
+
+    KDBindings::Signal<> error;
+
+    virtual int setTls(const File &cafile) = 0;
+    virtual int setUsernameAndPassword(const std::string &username, const std::string &password) = 0;
+    virtual int setWill(const std::string &topic, int payloadlen = 0, const void *payload = nullptr, int qos = 0, bool retain = false) = 0;
+
+    virtual int connect(const Url &host, int port = c_defaultPort, int keepalive = c_defaultKeepAliveSeconds) = 0;
+    virtual int disconnect() = 0;
+
+    virtual int publish(int *msgId, const char *topic, int payloadlen = 0, const void *payload = nullptr, int qos = 0, bool retain = false) = 0;
+
+    virtual int subscribe(const char *pattern, int qos = 0) = 0;
+    virtual int unsubscribe(const char *pattern) = 0;
+};
+
+/*
+ * Class: MqttClient
+ *
+ * This class exposes mosquitto client functions to
+ * the application / business logic.
+ */
+class KDMQTT_API MqttClient : public IMqttClient
+{
+    friend class MqttUnitTestHarness;
+
+public:
+    MqttClient(const std::string &clientId, bool cleanSession = true, bool verbose = false);
```

Comment:
I'd rather turn the two bool parameters into a single `Flags` parameter for better readability on call site.